### PR TITLE
ASoC: SOF: refine suspend_ignored flag

### PR DESCRIPTION
**Pull request by keyonjie**
ASoC: SOF: refine suspend_ignored flag

Move the flag from snd_sof_pcm_stream to snd_sof_dev, use count to store
the number of d0i3 compatible streams being suspended, and then we don't
need the helper snd_sof_dsp_d0i3_on_suspend() anymore.

### CODE REVIEW

**Comment by ranj063** [comment]
@keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?

**Comment by keyonjie** [comment]
> @keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the 

Yes, it came from yours, sorry about missing your authorship here as I split it from the big local change here, correcting it now. 

> second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?

Yes, your 2nd commit goes further, but I want it to go even further when with S0 D0I3 support enabled, all these are already done on my side, I am refining remained local changes(there might be 5~8 commits)  and will submit them soon.

**Comment by ranj063** [comment]
> > @keyonjie this PR does exactly the same as #1481 but with a different name for the count. But the
> 
> Yes, it came from yours, sorry about missing your authorship here as I split it from the big local change here, correcting it now.
> 
> > second commit in my PR goes one step further in actually simplifying the sequence too. Why is this one better ?
> 
> Yes, your 2nd commit goes further, but I want it to go even further when with S0 D0I3 support enabled, all these are already done on my side, I am refining remained local changes(there might be 5~8 commits) and will submit them soon.

@keyonjie dont worry about authorship. not a big deal, we're the same team. lets put all patches in one PR. It will be hard to see the entire picture if you split them up.

**Comment by keyonjie** [comment]
@ranj063 @plbossart OK, I will submit S0 D0i3 support commits here together.

**Comment by keyonjie** [comment]
Closing this, it will be included in #1524 soon.

**Comment by keyonjie** [comment]
@ranj063 @plbossart S0 D0I3 is implemented now, please help review.

File: sound/soc/sof/sof-priv.h [hunk 1]
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
...
+#define	SOF_DSP_STATE_BIT_I3		BIT(1) /* dsp is in I3 or I0 */
+/* dsp power state is under updating */
+#define	SOF_DSP_STATE_BIT_UPDATE	BIT(2)
+
+#define	SOF_DSP_STATE_D3	0 /* dsp is in D3 */
+#define	SOF_DSP_STATE_D0	SOF_DSP_STATE_BIT_D0 /* dsp is in D0 */
+/* dsp is in D0I3(low power state) */
+#define	SOF_DSP_STATE_D0I3	(SOF_DSP_STATE_BIT_D0 | SOF_DSP_STATE_BIT_I3)
+/* dsp will update to either D0 or D0I3 state automatically */
+#define	SOF_DSP_STATE_AUTO	SOF_DSP_STATE_BIT_UPDATE
```
Comment (ranj063) [inline sound/soc/sof/sof-priv.h]: @keyonjie this patch is just difficult to understand "ASoC: SOF: refine DSP power states". Care to break it down?

File: sound/soc/sof/sof-priv.h [hunk 2]
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
 	(IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE) || \
 	 IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST))
 
-/* DSP D0ix sub-state */
-enum sof_d0_substate {
-	SOF_DSP_D0I0 = 0,	/* DSP default D0 substate */
-	SOF_DSP_D0I3,		/* DSP D0i3(low power) substate*/
-};
+#define	SOF_DSP_STATE_BIT_D0		BIT(0) /* dsp is in D0 or D3 */
```
Comment (ranj063) [inline sound/soc/sof/sof-priv.h:60]: the comment makes no sense. Is the DSP in D0 or D3?

File: sound/soc/sof/sof-priv.h [hunk 3]
```diff
@@ -54,11 +54,17 @@ extern int sof_core_debug;
...
+#define	SOF_DSP_STATE_BIT_I3		BIT(1) /* dsp is in I3 or I0 */
+/* dsp power state is under updating */
+#define	SOF_DSP_STATE_BIT_UPDATE	BIT(2)
```
Comment (ranj063) [inline sound/soc/sof/sof-priv.h:63]: I am not a big fan of bit masks for dsp state. Its not like you could have 2 bits  set at the same time. Enum is a lot more intuitive.

File: sound/soc/sof/pm.c [hunk 4]
```diff
@@ -287,7 +264,7 @@ int snd_sof_suspend(struct device *dev)
 	if (sdev->d0i3_suspend_cnt) {
 		/* suspend to D0i3 */
 		dev_dbg(sdev->dev, "DSP is trying to enter D0i3...\n");
-		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I3);
+		ret = snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_D0I3);
```
Comment (ranj063) [inline sound/soc/sof/pm.c]: I think you should move this to snd_sof_dsp_suspend() as well.

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -152,15 +153,21 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	/* get runtime PCM params using widget's stream name */
 	spcm = snd_sof_find_spcm_name(scomp, swidget->widget->sname);
 	if (!spcm) {
-		dev_err(scomp->dev, "error: cannot find PCM for %s\n",
+		dev_err(sdev->dev, "error: cannot find PCM for %s\n",
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: please dont change this. let it be scomp->dev

File: sound/soc/sof/pm.c [hunk 6]
```diff
@@ -258,7 +235,7 @@ int snd_sof_resume(struct device *dev)
 	if (sdev->d0i3_suspend_cnt) {
 		/* resume from D0I3 */
 		dev_dbg(sdev->dev, "DSP will exit from D0i3...\n");
-		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I0);
+		ret = snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_D0);
```
Comment (ranj063) [inline sound/soc/sof/pm.c]: same here. should be moved to snd_sof_dsp_resume()

File: sound/soc/sof/intel/hda.c [hunk 7]
```diff
@@ -635,6 +639,8 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 #endif
 	hda_codec_i915_exit(sdev);
 
+	hda_dsp_power_state_reset(hda);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:675]: @keyonjie this looks incorrect. Typically when you remove the SOF device, you resume it and bring it back to full power. So the DSP cant be in D3. 

Besides reset is hardly intuitive to convey the message that the DSP is in D3. please change the name.

File: sound/soc/sof/intel/hda.h [hunk 8]
```diff
@@ -409,6 +411,8 @@ struct sof_intel_hda_dev {
 	int irq;
 
 	/* PM related */
+	u32 dsp_power_state;
+	struct mutex ps_mutex;	/* protects dsp_power_state change */
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.h:416]: I am also not able to fully appreciate the need for this mutex, at least just yet while reviewing only the second patch of the series. We only set the power states during suspend/resume and they can never overlap

File: sound/soc/sof/intel/hda-pcm.c [hunk 9]
```diff
@@ -213,10 +213,10 @@ snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
 }
 
 int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
-		     struct snd_pcm_substream *substream)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-pcm.c:216]: @keyonjie which "SOF specific information" do you need easy access to. Neither the patch nor the commit message make it clear

File: sound/soc/sof/intel/hda-dsp.c [hunk 10]
```diff
@@ -446,16 +449,41 @@ int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
...
 }
 
+void hda_dsp_d0i3_work(struct work_struct *work)
+{
+	struct sof_intel_hda_dev *hda =
+		container_of(work, struct sof_intel_hda_dev,
+			     d0i3_work.work);
+	struct snd_sof_dev *sdev = dev_get_drvdata(hda->dev);
+
+	/* keep DMA trace for delay D0I3 */
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c]: "delay D0i3" you mean D0i3 in S0?

File: sound/soc/sof/intel/hda-pcm.c [hunk 11]
```diff
@@ -227,12 +228,21 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 
 	/* binding pcm substream to hda stream */
 	sstream->substream->runtime->private_data = &dsp_stream->hstream;
+
+	/* update active pcm counts */
+	hda->pcm_active++;
+	if (sstream->d0i3_compatible)
+		hda->lp_pcm_active++;
+
+	snd_sof_dsp_set_power_state(sdev, SOF_DSP_STATE_AUTO);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-pcm.c]: Do you really need this here? You can do this after IPC TX no?

File: sound/soc/sof/intel/hda.c [hunk 12]
```diff
@@ -597,6 +600,9 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 	struct pci_dev *pci = to_pci_dev(sdev->dev);
 	const struct sof_intel_dsp_desc *chip = hda->desc;
 
+	/* First Cancel any pending attempt to put DSP to D0i3 */
+	cancel_delayed_work_sync(&hda->d0i3_work);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c:632]: don't you also need to cancel this work during system suspend?

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h:63]: Plus don't invent new definitions. I3 does not exist, it's D0i3.

File: sound/soc/sof/intel/hda-dsp.c [hunk 13]
```diff
@@ -338,49 +338,154 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
...
+
+	/* success, update power state */
+	hda->dsp_power_state = SOF_DSP_STATE_D0;
 
+exit:
+	mutex_unlock(&hda->ps_mutex);
 	return ret;
 }
 
+int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:398]: There's too  much in common between set D0i0 and set_D0i3. I think we can unify these 2 functions.

File: sound/soc/sof/intel/hda-dsp.c [hunk 14]
```diff
@@ -334,8 +334,9 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
 	pm_gate.flags = flags;
 
 	/* send pm_gate ipc to dsp */
-	return sof_ipc_tx_message(sdev->ipc, pm_gate.hdr.cmd, &pm_gate,
-				  sizeof(pm_gate), &reply, sizeof(reply));
+	return sof_ipc_tx_message_nopm(sdev->ipc, pm_gate.hdr.cmd,
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:337]: is this change really necessary? You'd always send the pm_date_ipc before the DSp is put in D0i3, isnt it?

**Comment by keyonjie** [comment]
WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).

I prefer to update to a new version firstly next week when I am back to Shanghai.

**Comment by plbossart** [comment]
> WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).
> 
> I prefer to update to a new version firstly next week when I am back to Shanghai.

I don't mean to sound rude but if you have 200+ comments you probably need to recheck this PR on your own and resubmit when it's mature.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/topology.c]: oops, I don't know why I changed this, fix it now.

File: sound/soc/sof/intel/hda-dsp.c (hunk 13 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c:398]: um, 3 biggest differences that why I split them, the 1st is that we only need delayed work for set D0I3(not for set D0I0), the 2nd one is that we need opposite sequence of IPC sending and register setting, and the 3rd one is that we need vary POWER_GATE IPC flags for D0I3 only.

File: sound/soc/sof/sof-priv.h (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h]: sure, let me move these macro definitions to a separated patch.

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h:63]: @ranj063 @plbossart let me do this in another way in the coming update version.

File: sound/soc/sof/sof-priv.h (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/sof-priv.h:60]: maybe like this?
/* 1 -- dsp is in D0, 0 -- dsp is in D3 */

File: sound/soc/sof/pm.c (hunk 6 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c]: Yes, actually.

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.h:416]: they could be set in IPC sending and PCM open/close also in the subsequent patches.

File: sound/soc/sof/intel/hda.c (hunk 7 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c:675]: > @keyonjie this looks incorrect. Typically when you remove the SOF device, you resume it and bring it back to full power. So the DSP cant be in D3.

Really? Why we need to keep it in full power if it is not used?

> 
> Besides reset is hardly intuitive to convey the message that the DSP is in D3. please change the name.

@ranj063 what name do you suggest to use? My understanding "reset to xxx state" is quite common?

File: sound/soc/sof/intel/hda-dsp.c (hunk 10 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c]: Yes, let me change the comment.

File: sound/soc/sof/intel/hda-pcm.c (hunk 11 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-pcm.c]: Here is PCM open, Do you mean doing it after trigger start IPC, in pcm_trigger?

File: sound/soc/sof/intel/hda.c (hunk 12 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda.c:632]: you mean D0I3 suspend? that is already covered by hda_dsp_suspend()->hda_dsp_set_power_state() here:
```
	case SOF_DSP_STATE_D0I3:
		/* First Cancel any pending attempt to put DSP to D0i3 */
		cancel_delayed_work_sync(&hda->d0i3_work);

		/* disable DMA trace for no delay D0I3 */
		ret = hda_dsp_set_d0i3(sdev, HDA_PM_NO_DMA_TRACE);

```

File: sound/soc/sof/intel/hda-dsp.c (hunk 14 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c:337]: It is needed, without this x_nopm() helpers, it will wake up the DSP to D0I0 before sending the pm_gate_ipc.

**Comment by keyonjie** [comment]
> > WoW, looks this will become another 200+ comments PR easily, if I reply them one by one, :).
> > I prefer to update to a new version firstly next week when I am back to Shanghai.
> 
> I don't mean to sound rude but if you have 200+ comments you probably need to recheck this PR on your own and resubmit when it's mature.

That's true, thanks for reminding.

**Comment by keyonjie** [comment]
@plbossart @ranj063 Happy thanks giving day!

I think I addressed most of the comments in the updated version, hope you like this version.

I talked with @RanderWang and realize that we might need to change to put Audio DSP to D0I3 for runtime suspend, but I think there is no conflict with what I implemented here, though it might be possible to combine these 2 scenarios with change in ASoC PCM handling(I did this try at the very beginning of my start of the D0Ix task in https://github.com/thesofproject/linux/pull/1154), some people might not like this solution as it change the component device RT PM logic in soc-dpcm.c when most of their platform actually don't need that extra logic.

File: sound/soc/sof/intel/hda-dsp.c [hunk 15]
```diff
@@ -338,49 +338,155 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
...
+
+	dev_vdbg(bus->dev, "D0I3C updated, register = 0x%x\n",
+		 snd_hdac_chip_readb(bus, VS_D0I3C));
+
+	hda->dsp_power_state = SOF_DSP_STATE_D0I3;
+
+exit:
+	mutex_unlock(&hda->ps_mutex);
+	return ret;
+}
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dsp.c:451]: @keyonjie I'm going to leave a long comment about the changes in this commit.

1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume. 

2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

File: sound/soc/sof/intel/hda.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda.c:675]: @keyonjie I dont know in my opinion, the mutex isnt required, so just set the power state directly here and remove the init/reset functions

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda.h:416]: @keyonjie that was my other comment, why do we need to do it in PCM open/close, can we not do it just when we send the IPC?

File: sound/soc/sof/pm.c [hunk 16]
```diff
@@ -255,7 +255,7 @@ int snd_sof_resume(struct device *dev)
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	int ret;
 
-	if (snd_sof_dsp_d0i3_on_suspend(sdev)) {
+	if (sdev->d0i3_suspend_cnt) {
```
Comment (plbossart) [inline sound/soc/sof/pm.c]: This direction is not quite right.
It turns out that for SoundWire we do have to keep the DSP powered if we want to avoid complete re-enumeration when exiting clock stop. This could happen in the complete absence of 'D0i3 compatible' streams.
this is really missing an idle loop detection that is based on 
a) D0i3-compatible streams being opened
b) no IPC for a while
c) the state of child devices, e.g. if a SoundWire master remains 'active' preventing pm_runtime entry.

Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart I didnt quite understand this ask. The change here only really affects system suspend when there are D0i3-compatible streams running in the DSP. Are you saying that we also need to account for the fact that the SDW master could be active?

Comment (plbossart) [reply sound/soc/sof/pm.c]: what I am saying is that while doing a system suspend, the device may already be in D0i3 and remain there (S0 target) or move to D3 (S3 target).

The use of refcounts would be quite misleading here, it's only one of the reasons why the device might remain in D0i3.

Comment (ranj063) [reply sound/soc/sof/pm.c]: @plbossart agreed wrt the WoV usecase this is the only reason that the DSP would remain in D0i3. Maybe we're missing the line of sight for the other possible indicators for this. Can we possibly think of adding this in a separate PR?

Comment (plbossart) [reply sound/soc/sof/pm.c]: well no. You remove a criterion to replace it by a refcount, we are not going to modify this again later to come back to another criterion. We need to respect the time of reviewers...

Comment (keyonjie) [reply sound/soc/sof/pm.c]: @plbossart I think I get your point here, I talked with @RanderWang about keeping the DSP in D0I3 for runtime suspend (for SDW clock stop supporting), I did't intend mix that with the D0I3-compatible stream running on D0I3, as I commented in https://github.com/thesofproject/linux/pull/1524#issuecomment-559761166.

Comment (keyonjie) [reply sound/soc/sof/pm.c]: Hi @plbossart @ranj063 so the state machine could be updated like this:

```
/*
 * Audio DSP states may transform as below:+
 *
 *                +---------------------+  runtime resume
 *                |                     <---------------------+
 *   +------------>                     |  runtime suspend    |
 *   |            |       D0(active)    +----------------+    |
 *   |            |                     +------------+   |    |
 *   |       +----+                     |    D0I3    |   |    |
 *   |       |    +-------^----+----^-- +  compatible|   |    |
 *   |       |            |    |    |        stream  |   |    |
 *   |resume |            |    |    |        opened  |   |    |
 *   | from  |  D3        |    |    |        only    |   |    |
 *   |  D3   |suspend     |    |    |                |   |    |
 *   |       |            |    |    |The last        |   |    |
 *   |       |            |    |    |opened D0i3     |   |    |
 *   |       |            |    |    |compatible      |   |    |
 *   |       |            |    |    |stream closed   |   |    |
 * +-+-------v-------+    |    |    |            +---v---v----+---+
 * |                 |    |    |    +------------+                |
 * |  D3 (suspended) |    |    | D0I3 suspend    |      D0I3      |
 * |                 |    |    +----------------->                +---+
 * |                 |    +----------------------+                |   |
 * +-------^---------+         resume from       +----------------+   |
 *         |                   D0I3 suspend                           |
 *         |                                   D3 suspend             |
 *         +----------------------------------------------------------+
 *
 * d0i3_suspend = s0_suspend && D0I3 stream opened,
 * D3 suspend = !d0i3_suspend,
 */

```

Comment (ranj063) [reply sound/soc/sof/pm.c]: @keyonjie are you saying that you'd never enter D3 during runtime suspend then>?

File: sound/soc/sof/intel/hda-dsp.c (hunk 15 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-dsp.c:451]: > @keyonjie I'm going to leave a long comment about the changes in this commit.
> 
> 1. The checks for power state inside the set_D0i0/set_D0i3 are a bit of an overkill. In reality, there are only 2 options: enter D0i3 or exit D0i3. Entering is always from D0, you'd never really see the case where you're trying to enter D0i3 from D3. Likewise, exiting D0i3 is only possible if you're already in D0i3. Otherwise, it would be a system resume or a runtime resume.

That's true, but since the enum dsp_power_state can have the D3 value, we only implement the helpers here, we can't make sure the caller will use it correctly always?

Safe check is a compromise, some scenarios might never happen(e.g. D3->D0I3 entry, D3->D0I3 exit), some might happen via error used(e.g. D0I3->D0I3 entry, D0->D0I3 exit).

I preferred to add these check here to avoid critical errors happen or disordering the state(e.g. if we do try doing D3->D0I3), but I am fine if both you and @plbossart think we should remove the D3 check.

> 2. I still do not appreciate the mutex. The commands to enter/exit D0i3 are already protected by waiting for the CIP timeout. Do we really need a mutex for just setting the hda->dsp_power_state? This is not really a register write. Maybe @kv2019i and @plbossart can correct me if I am wrong

I introduced the mutex as I think we could have race in setting and getting this dsp_power_state, especially considering that the hda_dsp_set_d0i3() could be a delayed work which could run in another thread.

> 3. The bit masks for the power states are also not pretty. There are 3 states, why cant we just simply use an enum instead of setting a D0 bit along with I3 bit? Makes it a lot easier to understand

The key point here is that sometimes the caller(e.g. after a normal IPC sent) actually don't know which state it should put the DSP to, I introduce a command SOF_DSP_STATE_SET_AUTO here for these scenarios. So the rule here is only using command when calling hda_dsp_set_power_state(), and leave the DSP state change decision to the dsp state management logic(inside hda-dsp.c) only.

```
enum sof_dsp_state {
	SOF_DSP_STATE_D3 = 0,
	SOF_DSP_STATE_D0 = SOF_DSP_STATE_BIT_D0,
	SOF_DSP_STATE_D0I3 = SOF_DSP_STATE_BIT_D0 | SOF_DSP_STATE_BIT_I3,
};

enum sof_dsp_state_cmd {
	SOF_DSP_STATE_SET_D3 = SOF_DSP_STATE_D3,
	SOF_DSP_STATE_SET_D0 = SOF_DSP_STATE_D0,
	SOF_DSP_STATE_SET_D0I3 = SOF_DSP_STATE_D0I3,
	SOF_DSP_STATE_SET_AUTO = SOF_DSP_STATE_BIT_UPDATE,
};
```

File: sound/soc/sof/pm.c (hunk 16 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c]: Yes, @ranj063 enter to D0I3 only for runtime suspend, if the platform doesn't support D0I3, the power gating will still work when it is not used, for power saving.

File: sound/soc/sof/pm.c [hunk 17]
```diff
@@ -284,7 +284,7 @@ int snd_sof_suspend(struct device *dev)
...
```
Comment (kv2019i) [inline sound/soc/sof/pm.c]: Hey, I wasn't on the review list, so I'm a bit late to this. I'm not thrilled about adding a refcount, especially as the places in code where it is incremented and decremented, have a lot of additional  conditional logic of their own. When/if something goes wrong, it seems painful to debug where the refcount went wrong. With old snd_sof_dsp_d0i3_on_suspend(), you at least have a single place to check which stream is in unexpected state. And a place to add new logic that can keep DSP at D0i3.

File: sound/soc/sof/intel/hda.h (hunk 8 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda.h:416]: @keyonjie This is not a simple lock. In the comment this indicates the lock protects "dsp_power_state", but reading the code, the lock actually covers DSP power state as well. E.g. if you change the DSP power state, you need to hold this lock. Do we really need, or is a lock protecting just the dsp_power_state enough (and if yes, do we need a lock at all)?

File: sound/soc/sof/pm.c (hunk 16 above)
Comment (plbossart) [reply sound/soc/sof/pm.c]: NAK, runtime suspend and D0i3 are mutually incompatible.

File: sound/soc/sof/pm.c (hunk 17 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c]: > Late to the review, but a few specific comments (on new locks and refcounts), please see inline.

Thanks for review @kv2019i, it comes in time we are still in discussion and I just get myself free from other higher priority issues.

Comment (keyonjie) [reply sound/soc/sof/pm.c]: > Hey, I wasn't on the review list, so I'm a bit late to this. I'm not thrilled about adding a refcount, especially as the places in code where it is incremented and decremented, have a lot of additional conditional logic of their own. When/if something goes wrong, it seems painful to debug where the refcount went wrong. With old snd_sof_dsp_d0i3_on_suspend(), you at least have a single place to check which stream is in unexpected state. And a place to add new logic that can keep DSP at D0i3.

Yes, @plbossart also mentioned about this, my opinion is that if the criteria we can foresee is only that "the count of D0I3 compatible streams being suspend > 0", I agree with @ranj063 that using this simple refcount is fine, but if we can foresee that we need to add some other criteria (e.g. for SDW) for D0I3 entry, yes, I agree with you and @plbossart here.

**Comment by keyonjie** [comment]
> @keyonjie This is not a simple lock. In the comment this indicates the lock protects "dsp_power_state", but reading the code, the lock actually covers DSP power state as well. E.g. if you change the DSP power state, you need to hold this lock. Do we really need, or is a lock protecting just the dsp_power_state enough (and if yes, do we need a lock at all)?

Yes, the intention is to protect the whold DSP power state change process.

**Comment by keyonjie** [comment]
> Late to the review, but a few specific comments (on new locks and refcounts), please see inline.

Thanks for review @kv2019i, it comes in time we are still in discussion and I just get myself free from other higher priority issues.

File: sound/soc/sof/pm.c (hunk 16 above)
Comment (keyonjie) [reply sound/soc/sof/pm.c]: @plbossart can you help guide how the state machine should look like in your brain?

**Comment by keyonjie** [comment]
@plbossart @ranj063 @kv2019i the PR is updated with removing adding the d0i3_suspend_cnt(and keep using the helper snd_sof_dsp_d0i3_on_suspend() which can be extended for SDW later), I keep the usage of the mutex to protect the DSP power state access *and the state changing itself* as the state change can take long time during the delayed_work, e.g. if we forget to cancel the delayed work before trying to clear/reset the state, this mutex will help to reject this kind of operation.

**Comment by plbossart** [comment]
> @plbossart @ranj063 @kv2019i the PR is updated with removing adding the d0i3_suspend_cnt(and keep using the helper snd_sof_dsp_d0i3_on_suspend() which can be extended for SDW later), I keep the usage of the mutex to protect the DSP power state access _and the state changing itself_ as the state change can take long time during the delayed_work, e.g. if we forget to cancel the delayed work before trying to clear/reset the state, this mutex will help to reject this kind of operation.

@keyonjie I am not going to review 2 PRs for the same things. Please work with @ranj063 and suggest a common solution.

**Comment by keyonjie** [comment]
> @keyonjie I am not going to review 2 PRs for the same things. Please work with @ranj063 and suggest a common solution.

Sorry for the confusion, yes, please hold on for a moment, will let you know once we are aligned.

**Comment by keyonjie** [comment]
#1631 will be the replacement for this, closing this one.
